SIPS/sip-20.md

```diff
@@ -1,189 +1,237 @@
+type GetWebSocketsResult = WebSocketConnection[];
+```
+
+An example of usage is given below.
 
-The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+```typescript
+snap.request({
+  method: "snap_getWebSockets",
+})
+```
 
-### Snap implementation
 
-This SIP introduces a new `onExternalData` function export that MAY be implemented by the Snap. The function is called every time the WebSocket client receives some data. The function SHOULD accept an object parameter that MUST have `data`, `type` and `source` properties.
-The parameters represent the following:
-- `data`: An object that represents the data that was received. Depending on the `data.type`, this SHOULD contain a `message` property, which SHOULD be either a `string` or an `ArrayBuffer`.
-- `type`: The type of the external data source. For this SIP, this SHOULD always be `websocket`.
-- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints. This SHOULD be the exact same URL as specified in manifest.
+### Handling WebSocket Events
 
-The specification of types of `onExternalData` and its parameters:
+Snaps can handle WebSocket events by implementing the `onWebSocketEvent` handler. This handler will be called whenever a WebSocket event occurs, such as receiving a message, opening a connection, closing a connection, or encountering an error.
 
 ```typescript
-/**
- * A text message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as a string.
- */
-type WebSocketTextMessage = {
-  type: 'text';
-  message: string;
+import { OnWebSocketEventHandler } from "@metamask/snap-sdk";
+
+export const onWebSocketEvent: OnWebSocketEventHandler = async ({ event }) => {
+  switch (event.type) {
+    case "message":
+      // Handle incoming message
+      console.log(`Message received from ${event.origin}:`, event.data);
+      break;
+    case "open":
+      // Handle connection opened
+      console.log(`WebSocket connection opened with ID ${event.id} from ${event.origin}`);
+      break;
+    case "close":
+      // Handle connection closed
+      console.log(`WebSocket connection closed with ID ${event.id} from ${event.origin}`);
+      break;
+    case "error":
+      // Handle error
+      console.error(`WebSocket error occurred with ID ${event.id} from ${event.origin}`);
+      break;
+  }
 };
+```
+the type for an `onWebSocketEvent` handler function's arguments is:
 
-/**
- * A binary message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as an ArrayBuffer.
- */
-type WebSocketBinaryMessage = {
-  type: 'binary';
-  message: ArrayBuffer;
+```typescript
+export type WebSocketMessageEvent = {
+  type: "message";
+  id: string;
+  origin: string;
+  dataType: "text" | "binary";
+  data: string | Uint8Array;
```

Comment:
Ideally we use a discriminated union here to determine the type of `data`, like in the previous specification.